### PR TITLE
cloudwatch: Add resource_arns template query function

### LINKED ISSUE #8207: [Feature request] (cloudwatch) new templating function which use GetResources()
I think this is useful for RDS/ElastiCacher user.

The GetResources() API can filter resource by tags.
http://docs.aws.amazon.com/resourcegroupstagging/latest/APIReference/API_GetResources.html
By using this API, we can get ARN of resource which has `Enviroment=production` tag.

The templating function might becomes like following.
```
arn_filter_by_tag(region, resource_type, tags)
```

### PR DESCRIPTION
**Pull request by jeroenvollenbrock**
Fixes #8207 

Name | Description
------- | --------
*resource_arns(region, resource_type, tags)* | Returns a list of ARNs matching the specified `region`, `resource_type` and `tags`.

Query | Service
------- | -----
*resource_arns(eu-west-1,elasticloadbalancing:loadbalancer,{"elasticbeanstalk:environment-name":["myApp-dev","myApp-prod"]})* | ELB
*resource_arns(eu-west-1,ec2:instance,{"customer":["john-doe"]})* | EC2

### CODE REVIEW

**Comment by jeroenvollenbrock** [comment]
@torkelo Could this PR get a review please? 🙈

File: public/app/plugins/datasource/cloudwatch/datasource.ts [hunk 1]
```diff
@@ -293,6 +301,15 @@ export default class CloudWatchDatasource {
       return this.getEc2InstanceAttribute(region, targetAttributeName, filterJson);
     }
 
+
+    const resourceARNsQuery = query.match(/^resource_arns\(([^,]+?),\s?([^,]+?),\s?(.+?)\)/);
```
Comment (marefr) [inline public/app/plugins/datasource/cloudwatch/datasource.ts]: Would like a test of this. There's existing tests for metricFindQuery in datasource.test.ts so should be quite trivial to implement.

File: pkg/tsdb/cloudwatch/metric_find_query.go [hunk 2]
```diff
@@ -534,6 +537,65 @@ func (e *CloudWatchExecutor) handleGetEc2InstanceAttribute(ctx context.Context,
...
+		sess, err := session.NewSession(cfg)
+		if err != nil {
+			return fmt.Errorf("Failed to call ec2:NewSession, %v", err)
+		}
+		e.rgtaSvc = resourcegroupstaggingapi.New(sess, cfg)
+	}
+	return nil
+}
+
+func (e *CloudWatchExecutor) handleGetResourceArns(ctx context.Context, parameters *simplejson.Json, queryContext *tsdb.TsdbQuery) ([]suggestData, error) {
```
Comment (marefr) [inline pkg/tsdb/cloudwatch/metric_find_query.go]: Would like some tests of this to validate tjat expected input/output can be handled correctly.

File: pkg/tsdb/cloudwatch/metric_find_query.go [hunk 3]
```diff
@@ -585,6 +647,28 @@ func (e *CloudWatchExecutor) ec2DescribeInstances(region string, filters []*ec2.
 	return &resp, nil
 }
 
+func (e *CloudWatchExecutor) resourceGroupsGetResources(region string, filters []*resourcegroupstaggingapi.TagFilter, resourceTypes []*string) (*resourcegroupstaggingapi.GetResourcesOutput, error) {
```
Comment (marefr) [inline pkg/tsdb/cloudwatch/metric_find_query.go]: Would like some tests of this to validate tjat expected input/output can be handled correctly.

**Review by marefr — CHANGES_REQUESTED** [review]
Sorry for the time taken before reviewing this. In general I think this looks quite good. I would like you to add some tests for frontend and backend (see comments). Besides that I think we should be able to merge this.

**Comment by marefr** [comment]
@mtanda do you have time to have a look at this/test it? Thanks

**Comment by mtanda** [comment]
@marefr I tested it. The feature works well! 😄 

image

**Comment by marefr** [comment]
Thanks a lot @mtanda :+1:

File: public/app/plugins/datasource/cloudwatch/datasource.ts (hunk 1 above)
Comment (jeroenvollenbrock) [reply public/app/plugins/datasource/cloudwatch/datasource.ts]: Added a test using metricFindQuery in 3729660f5

File: pkg/tsdb/cloudwatch/metric_find_query.go (hunk 2 above)
Comment (jeroenvollenbrock) [reply pkg/tsdb/cloudwatch/metric_find_query.go]: Added a test for handleGetResourceArns in 3729660f5

File: pkg/tsdb/cloudwatch/metric_find_query.go (hunk 3 above)
Comment (jeroenvollenbrock) [reply pkg/tsdb/cloudwatch/metric_find_query.go]: I did not add an explicit test for this method, as it is merely a helper function that combines paged AWS results into one big response and already gained full coverage through the test for handleGetResourceArns

**Comment by jeroenvollenbrock** [comment]
> Sorry for the time taken before reviewing this. In general I think this looks quite good. I would like you to add some tests for frontend and backend (see comments). Besides that I think we should be able to merge this.

Thank you for the review! I've added in the requested tests :)

**Review by marefr — APPROVED** [review]
Great job!

**Comment by marefr** [comment]
Thank you for contributing to Grafana!

**Pull request merged**